### ISSUE TITLE
Combining issue, with comment blocks

### ISSUE THREAD

**Issue by dazinator** [Shazwazza/Smidge#23]
I am combining `jquery.min.js` with `knockout-es5.js` in a bundle.

However something funny happens.

The last line of `jquery.min` is this: 

```
//# sourceMappingURL=jquery.min.map
```

The top of `knockout-es5.js` is this:

```
/*!
 * Knockout ES5 plugin - https://github.com/SteveSanderson/knockout-es5
 * Copyright (c) Steve Sanderson
 * MIT license
 */
```

However when `Smidge` produces the bundle file, it looks like this:

```
//# sourceMappingURL=jquery.min.map/*!
 * Knockout ES5 plugin - https://github.com/SteveSanderson/knockout-es5
 * Copyright (c) Steve Sanderson
 * MIT license
 */
```

The browser complains:

![image](https://cloud.githubusercontent.com/assets/3176632/13999049/4ce146e4-f132-11e5-9f29-9dc0f8183024.png)

This is probably because just appends to the previous script, but in this case i'm guessing a comment block is only valid at the start of a line?

I am using the latest code base to replicate this.

**Comment by dazinator** [comment]
And just to add some background, this problem happens when adding a file to bundle that starts with a  multi-line comment. knockoutjs and it's various plugins are good examples of scripts that start with multiline comments - even the minified ones seem to keep these comments in, so it doesn't appear to be that uncommon.

No idea why these venders produce minified files that keep the comments in  - would have thought they should be stripped out but hey ho.

**Comment by Shazwazza** [comment]
Cool will have a look, this is all an issue now cuz I've made a lot of changes to jsmin to fix a few other issues, I've got tons of tests written already for the fixes and changes and will add these cases too. If you need it to work now you can revert to the first jsmin commit of smidge

**Comment by dazinator** [comment]
Ah ok, no worries, it's ok, I can wait for the future release with this fix, i'll just keep js files that start with multi line comments "unbundled" for now.

**Comment by Shazwazza** [comment]
This should all be fixed, however JSMin will cause the current dist version of knockout: https://github.com/SteveSanderson/knockout-es5/blob/master/dist/knockout-es5.js to error because the code doesn't have semi-colons in the correct spot. For example, here: https://github.com/SteveSanderson/knockout-es5/blob/master/dist/knockout-es5.js#L604 that end of statement requires a semicolon because it's a declaration of a variable.

It seems that the committed (non dist) version of knockout doesn't have this problem.

I can 'fix' this by relaxing how JsMin deals with line breaks but it's less than ideal since it will leave a bit more whitespace then needed. I'm just going to test this with other libraries and see if there's issues.

**Comment by Shazwazza** [comment]
Ok, looks like there are too many libs with problems like this, i'll relax the JSMin a bit so it's not so strict with end-of statements.

I've now tested with lots of libs and all seems well with the js side of things.

**Comment by dazinator** [comment]
Ah yes I see what you mean.. You would think a decent linter would pick that sort of thing up with the missing semi colons.. I'll wait for the rc4 nuget release and then i'll try this out.

**Comment by Shazwazza** [comment]
Have just released :)

**Comment by dazinator** [comment]
Hmm.. still hitting this problem using RC4:

I attempted to bundle jquery.min, and knockout-es5 with the following result:

![image](https://cloud.githubusercontent.com/assets/3176632/14855046/0989a79c-0c8a-11e6-9403-573131f4035c.png)

**Comment by dazinator** [comment]
To get things to work I had to manually fix up the files for the time being:

1. Manually remove the source map declaration from the bottom of jquery, i.e this bit:

```
//# sourceMappingURL=jquery.min.map
```

I'll raise a related issue about sourcemaps - basically if smidge is processing the source code, I don't think keeping the original sourcemap is going to do much good.

2. Removed the multiline comment at the top of the knockous-es5.min.js script.

**Comment by Taritsyn** [comment]
Try to use the [JSMin.NET](https://github.com/Taritsyn/JSMin.NET) (.NET port of the Douglas Crockford's JSMin version of March 29, 2013).

**Comment by Shazwazza** [comment]
Hrm, this is strange. I have this libraries committed in the web demo project of Smidge and they combine/minify/etc... without error:

![image](https://cloud.githubusercontent.com/assets/1742685/14862369/55750d8c-0cb1-11e6-9332-b57a9e87edb9.png)

You can test this by cloning this project and pressing F5 to launch the test website.

Perhaps you didn't increment your smidge version parameter in your config and it's still using the exist cached bundle version?

**Comment by Shazwazza** [comment]
I suppose the other potential issue is that I bundled up RC4 incorrectly ... hrm. I'm also testing the font-awesome minification in the source and that also seems to work without issue.

**Comment by dazinator** [comment]
I am triple checking the cache, (deleted all the cache files just in case) and I am still hitting the same problem so I don't think its a caching thing.

I notice that are trying with jquery 1.12.2 and I wonder - does that version of jquery have the sourcemap declaration as the last line? 

I'm hitting this with jquery 2.1.4

**Comment by dazinator** [comment]
- When I get time, i'll try adding a unit test to replicate.

**Comment by Shazwazza** [comment]
no worries, i can also double check all of this later on too. Oh, and the jquery one committed does have the source map declaration as the last line.

**Comment by dazinator** [comment]
I just updated my fork so I am inline with latest master again. Interestingly all the unit tests pass, but if I try and start the tests site, I get this:

![image](https://cloud.githubusercontent.com/assets/3176632/14885222/473711f2-0d42-11e6-991e-4e1bb2a68a8c.png)

I'll step through the code and see if I can figure out the issue, but just thought I'd check to see if you had seen this before?

**Comment by dazinator** [comment]
I think I have spotted this issue..

JsMin is regsitered as a singleton instance:
```
 services.AddSingleton<IPreProcessor, JsMin>();
```

However when I start the site, I am seeing two threads / tasks do minification - both running against the same singleton instance. 

 Both threads enter JSMin and create an instance of a writer and store it in the same private field:

```

![image](https://cloud.githubusercontent.com/assets/3176632/14885451/5437eace-0d43-11e6-8d80-d6343c9f8211.png)

```

One thread then disposes that private instance, and the second thread can no longer write as its referencing the same private instance.

So basically this class isn't thread safe so registering it as a singleton is problematic.

Want me to raise a seperate issue?

**Comment by dazinator** [comment]
The issue of multipel threads sunning against JSMin seems to be sporadic. For example for some reason after x times of starting the site up (made absolutely no changes) I am now only seeing the one thread / task execute jsmin - so this thread safety issue goes away. 

After successfully starting the site, smidge created the bundles in the cache folder - which is cool. I decided to clear the folder, and then restart the site.

I am now seeing this:

![image](https://cloud.githubusercontent.com/assets/3176632/14885688/a5f28f62-0d44-11e6-953d-5cc1289af983.png)

By this point I have no idea what is going on.. I have not changed any of the source files, yet one time it was able to run fine and produce the bundle, and then clearing the cache and restarting, now decides to result in the above?

**Comment by Shazwazza** [comment]
I'm not sure why but this email thread doesn't seem to be
re-posting/updating the issue on GitHub ... oh well, i'll stick to email
for now.

Thanks for the info so far, i'll need to run some tests too... you are
probably also running via IIS whereas I'm running via DNX which may have
some affect. I'll double check the jsmin lifetime scope, might have been
something I overlooked or changed at some point. I should find time to
check all of this later today or tomorrow.

On Thu, 28 Apr 2016 at 14:26 Darrell <notifications@github.com> wrote:

> The issue of multipel threads sunning against JSMin seems to be sporadic.
> For example for some reason after x times of starting the site up (made
> absolutely no changes) I am now only seeing the one thread / task execute
> jsmin - so this thread safety issue goes away.
>
> After successfully starting the site, smidge created the bundles in the
> cache folder - which is cool. I decided to clear the folder, and then
> restart the site.
>
> I am now seeing this:
>
> [image: image]
> <https://cloud.githubusercontent.com/assets/3176632/14885688/a5f28f62-0d44-11e6-953d-5cc1289af983.png>
>
> By this point I have no idea what is going on.. I have not changed any of
> the source files, yet one time it was able to run fine and produce the
> bundle, and then clearing the cache and restarting, now decides to result
> in the above?
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly or view it on GitHub
> <https://github.com/Shazwazza/Smidge/issues/23#issuecomment-215407955>
>
-- 
-- 
Shannon Deminick
*w:* http://shazwazza.com
*e:* sdeminick@gmail.com
*t:* shazwazza

**Comment by dazinator** [comment]
Ok cheers, I have moved all this to its own issue: #31
